[Blazor][WebView] Abstractions

javiercn opened the pull request:
WIP This is the initial setup for the Blazor abstractions for running on a WebView. Just sharing the branch for now, details and a design write-up to come a bit later.

The goal is to provide the fundamental primitives for Blazor WebView as well as a "Headless" webview that we can use in our unit tests.

javiercn committed:
Initial project structure

javiercn committed:
tmp

javiercn committed:
tmp

javiercn committed:
tmp

javiercn committed:
Second checkpoint

javiercn committed:
Continue cleanups

javiercn commented:
@SteveSandersonMS @Eilon here's the branch with my WIP FYI, I'm not looking 4 feedback just yet, I'll ping the PR when I get this in the shape I want.

javiercn committed:
Cleanups

javiercn committed:
Make things build

SteveSandersonMS committed:
Move shared files to shared dir

SteveSandersonMS committed:
Minor tidying

javiercn committed:
Refactors based on feedback

javiercn committed:
Renames, comments, etc

javiercn committed:
Get rid of additional projects in favor of unit tests

javiercn committed:
Cleanup and setup tests

javiercn committed:
initial unit tests

SteveSandersonMS committed:
Try renaming to IpcSender/IpcReceiver

SteveSandersonMS committed:
IpcSender/IpcReceiver can't be scoped as they must exist before any page attaches

SteveSandersonMS committed:
Refactoring to explicit per-page context

SteveSandersonMS committed:
Get unit test back

SteveSandersonMS committed:
Test for rendering root component

SteveSandersonMS committed:
Make directory for services

SteveSandersonMS committed:
Add comment

SteveSandersonMS committed:
Factor out IPC serialization into a common type for reuse in tests

SteveSandersonMS committed:
Tidy usings

SteveSandersonMS committed:
Tidy up naming - Initialize -> AttachPage

SteveSandersonMS committed:
Validate that messages are on the expected origin

SteveSandersonMS committed:
Typo

SteveSandersonMS committed:
Delete redundant project

SteveSandersonMS committed:
Rename directory to match repo conventions

SteveSandersonMS committed:
Add projects for Webview2 and WPF

SteveSandersonMS committed:
Add empty project for WPF sample

SteveSandersonMS committed:
Make WPF sample able to run

SteveSandersonMS committed:
Reference WebView2 package

SteveSandersonMS committed:
Fix casing in WebView2

SteveSandersonMS committed:
Add WebView to WPF sample

SteveSandersonMS committed:
Supply IServiceProvider to BlazorWebView

SteveSandersonMS committed:
Add some static content

SteveSandersonMS committed:
Remove shared WebView2 platform since there's no commonality across WPF and WinForms (each has a separate WebView2)

SteveSandersonMS committed:
Begin startup sequence

SteveSandersonMS committed:
Receive web messages

Eilon commented on src/Components/WebView/Platforms/Wpf/src/BlazorWebView.cs:
```diff
@@ -0,0 +1,50 @@
+
+namespace Microsoft.AspNetCore.Components.WebView.Wpf
+{
+    public sealed class BlazorWebView : Control, IDisposable
```
I'll add a commit to apply my changes from https://github.com/dotnet/MobileBlazorBindings/pull/333 to here.

SteveSandersonMS replied:
Excellent - thanks!

SteveSandersonMS committed:
Handle async initialization

SteveSandersonMS committed:
Supply static content

SteveSandersonMS committed:
Slight tidyup

SteveSandersonMS committed:
Add comment

Eilon committed:
Improve WPF control startup

- I left some commented-out code in there with the expectation that it will eventually be needed. This way it shows the pattern of the properties that is expected in order for the control to function properly.

SteveSandersonMS commented:
@javiercn @Eilon I had to rebase this onto `main` (and hence force-push) to pick up all the recent TypeScript changes before doing new work in that area. Please be sure to get the latest version of this branch locally before doing any further work on it, otherwise you'll have some trouble with pushing commits later.

SteveSandersonMS committed:
Exclude WPF-related projects on non-Windows

SteveSandersonMS committed:
Fix unit tests

SteveSandersonMS committed:
Begin defining blazor.webview.js

SteveSandersonMS committed:
Fix typo

SteveSandersonMS committed:
Supply blazor.webview.js to browser

SteveSandersonMS committed:
Start making blazor.webview.js boot

javiercn commented:
> @javiercn @Eilon I had to rebase this onto `main` (and hence force-push) to pick up all the recent TypeScript changes before doing new work in that area. Please be sure to get the latest version of this branch locally before doing any further work on it, otherwise you'll have some trouble with pushing commits later.

Sounds good to me

javiercn commented on src/Components/WebView/WebView/src/Services/WebViewRenderer.cs:
```diff
@@ -0,0 +1,95 @@
+        private readonly Dictionary<string, int> _componentIdBySelector = new();
+        private readonly Dispatcher _dispatcher;
+        private readonly IpcSender _ipcSender;
+        private long nextRenderBatchId = 1;
```
Do we need this here?

For Blazor Server I can understand, but for Blazor WebView I think we can rely on a simple ACK without having to keep track of the id

SteveSandersonMS replied:
It just helps us discover more quickly if we have any bugs related to out-of-order delivery.

javiercn replied:
How can this situation arise?

SteveSandersonMS replied:
We can have bugs :)

Eilon replied:
BTW in my experience I've never seen such a bug while working on MBB and haven't seen it from customers, but perhaps it can more easily be an issue _while we develop_ the code, so it seems like a good precaution. I cannot imagine such a bug would be easy to debug when things simply explode and behave oddly.

javiercn commented on src/Components/WebView/WebView/src/IpcCommon.cs:
```diff
@@ -0,0 +1,71 @@
+            if (message != null && message.StartsWith(_ipcMessagePrefix, StringComparison.Ordinal))
+            {
+                var messageAfterPrefix = message.AsSpan(_ipcMessagePrefix.Length);
+                var parsed = (JsonElement[])JsonSerializer.Deserialize(messageAfterPrefix, typeof(JsonElement[]), JsonSerializerOptionsProvider.Options);
```
Something to think about here is that we are double-serializing JSON data when we do JS Interop

SteveSandersonMS replied:
Yeah, I agree. I think we can make this interop mechanism more low-level and performant later if we want to. Hopefully it's all very much contained to this one file (and the equivalent on the `.ts` side) so it should be easy to reason about any changes.

javiercn commented on src/Components/WebView/WebView/src/IpcReceiver.cs:
```diff
@@ -0,0 +1,119 @@
+        public IpcReceiver(WebViewManager manager)
+        {
+            _manager = manager;
+        }
```
I think we should reconsider the direction of this dependency.

The WebViewManager creates and manages the IPCReceiver, but then the receiver calls back to the manager and both reference each other.

I think we might be better of if we follow an alternative approach:
1) Handle initialization outside the IpcReceiver.
2) Instead of dispatching the "AttachToPage" we expose an "Initialize" event that the manager subscribes to from outside.

That way all interactions always go Manager -> (IPCXXX) and there is no coupling both ways.

SteveSandersonMS replied:
Absolutely, sounds good to me! I agree it's a bit unexpected that it relies on an `internal` method at the moment. The only drawback that concerns me about using an event is that it's hard to deal with asynchrony and exceptions, but we can figure something out.

SteveSandersonMS committed:
Add source map for blazor.webview.js

SteveSandersonMS committed:
Simplify

SteveSandersonMS committed:
Treat as secure origin to enable more JS APIs

SteveSandersonMS committed:
Begin enabling actual IPC both ways

SteveSandersonMS committed:
Actually renders a component

SteveSandersonMS committed:
Basic event dispatching

SteveSandersonMS committed:
Reduce lag

SteveSandersonMS committed:
Allow platform to observe startup duration

Eilon committed:
Move non-WPF specific WebView2 code into common assembly

This will enable WinForms to re-use 95% of the logic.

Eilon commented on src/Components/WebView/Platforms/WebView2/src/WebView2WebViewManager.cs:
```diff
@@ -0,0 +1,72 @@
+
+namespace Microsoft.AspNetCore.Components.WebView.WebView2
+{
+    public class WebView2WebViewManager : WebViewManager
```
At some point we'll need to decide whether we want all this stuff to be public. It's not really an interesting part of the API surface from a public consumption perspective, yet it allows us to have common code between WPF, WinForms, and .NET MAUI.

No need to decide anything now. Keeping this in a separate assembly will help us ensure we have good layering for now.

SteveSandersonMS replied:
I agree, the layering is fine for now.

Later on we should consider replacing the M.A.C.WebView.WebView2 package with use of a shared-source file for the `WebView2WebViewManager` abstraction. It would make no difference to functionality, but would mean we just don't have to ship an extra publicly-consumable package for what is intended as an internal abstraction.

SteveSandersonMS committed:
Add some XML docs

SteveSandersonMS committed:
Prevent multi-initialization

SteveSandersonMS committed:
Spare platform-specific controls the burden of having to keep track of whether a page is currently attached

Let them add/remove root components at any time

SteveSandersonMS committed:
Fix unit tests

SteveSandersonMS committed:
Let WPF control specify its own root components via bindable collection

SteveSandersonMS committed:
Render an actual .razor component (via compiler workaround)

SteveSandersonMS committed:
Let the WPF developer specify the host page

campersau commented on src/Components/WebView/Platforms/Wpf/src/BlazorWebView.cs:
```diff
@@ -0,0 +1,150 @@
+                // Dispatch because this is going to be async, and we want to catch any errors
+                WpfDispatcher.Instance.InvokeAsync(async () =>
+                {
+                    var newItems = eventArgs.OldItems.Cast<RootComponent>();
+                    var oldItems = eventArgs.NewItems.Cast<RootComponent>();
```
Shouln't this be like this?

```suggestion
                    var newItems = eventArgs.NewItems.Cast<RootComponent>();
                    var oldItems = eventArgs.OldItems.Cast<RootComponent>();
```

Also both `foreach` loops are doing `newItems.Except(oldItems)` shloudn't one be the other way around?

SteveSandersonMS replied:
Well spotted :) Fixed.

SteveSandersonMS committed:
Dispose on application exit

SteveSandersonMS committed:
Fix bug

SteveSandersonMS committed:
Clean up IPC code in TypeScript

SteveSandersonMS committed:
Notify about unhandled exceptions

SteveSandersonMS committed:
Support navigation and JS interop

SteveSandersonMS committed:
Support reloading when not on the index.html page

javiercn commented on src/Components/WebView/WebView/src/WebViewManager.cs:
```diff
@@ -0,0 +1,184 @@
+                await _currentPageContext.Renderer.AddRootComponentAsync(
+                    rootComponent.ComponentType,
+                    selector,
+                    rootComponent.Parameters);
```
I'm not sure about automatically re-rendering the components in this situation.

There's a chance that rootComponent.Parameters contains objects that have been mutated and given that we might go to a different host page, there is a chance that the same root components might not apply to the new page?

SteveSandersonMS replied:
Do you have more of an end-to-end scenario in mind that would be a good example of this?

I would expect in most cases people won't use the root-component parameters at all, and in the cases where they do, it's so that different host pages can have different global settings, or so that you could have a single root component shared across Blazor WebAssembly/Server/WebView and pass them some info or flags that differ based on the hosting environment. In these scenarios it wouldn't be something you'd try to mutate, and if you did try and nothing happened, that seems like it would be fine.

> given that we might go to a different host page

As it stands, the `BlazorWebView` controls only let you configure a single host page per control, so the scenario doesn't apply.

javiercn commented on src/Components/WebView/WebView/src/WebViewManager.cs:
```diff
@@ -0,0 +1,184 @@
+        protected bool TryGetResponseContent(string uri, bool isPageLoad, out int statusCode, out string statusMessage, out Stream content, out string headers)
+            => _staticContentProvider.TryGetResponseContent(uri, isPageLoad, out statusCode, out statusMessage, out content, out headers);
+
+        internal async Task AttachToPageAsync(string baseUrl, string startUrl)
```
Also, "spicy" thought is that we need to have a "root component" SetParameters API here to dispatch parameter updates to top-level components.

SteveSandersonMS replied:
Yes, if this becomes a common scenario then I definitely agree. However the usability of mutating root component parameters from a WPF/WinForms control might not be great, and there will always be alternatives like sending commands via some DI service. I'd suggest we wait to see some specific customer demand here before implementing this.

Eilon committed:
Fix filename

Eilon committed:
Add WinForms BlazorWebView and sample

Should be equivalent functionality and limitations to WPF.

SteveSandersonMS committed:
Experiment with tightening the API for adding root components

SteveSandersonMS committed:
Apply better defaults for desktop apps

SteveSandersonMS committed:
Don't log a bogus 404 error in the dev tools console

SteveSandersonMS commented:
@javiercn @Eilon We should try to decide what milestone we're trying to reach before merging this into `main`. There's a lot of worthy functionality in here already, so I'd say it would be good to merge soon.

What are the remaining pieces we want to have in before merging? The main one that comes to mind for me is more tests. There's opportunity to backfill a load of unit tests for `WebViewManager`, plus @javiercn is looking into E2E automation. Anything else?

SteveSandersonMS committed:
More build exclusions

javiercn committed:
* Implement Dispose pattern in WebViewManager
* Layering cleanups
* Tests and bug fixes

Eilon commented:
@SteveSandersonMS spot on. BTW for WinForms I took the lazy route and used your RootComponent type. Certainly not meant to be a final API, so thanks for updating it!

Regarding merging into main: Let's discuss today and we should easily be able to assign tasks to get them done, especially the more trivial ones (headers, doc comments, etc.).

Eilon committed:
Fixing another bad Web2View name

Eilon committed:
Make tests compile

Eilon committed:
Add a test

javiercn committed:
Add public API baselines, cleanup, fix build issues

Eilon committed:
Add license headers

Eilon committed:
Try to fix CI failure for WebView2Loader during Pack

javiercn committed:
Add missing doc comments

javiercn committed:
Fix npmproj

dougbu commented:
> any concerns we need to address about adding Windows-specific projects to the build

We have a few projects that can only be built in `msbuild` e.g. https://github.com/dotnet/aspnetcore/pull/30456/files#diff-366b25472963c6b33188d84139f6c8997298c5adc857a2624ab0d26a2d87baf0R9-R14 and a few others that are platform-specific. As long as the projects listed in https://github.com/dotnet/aspnetcore/pull/30456/files#diff-366b25472963c6b33188d84139f6c8997298c5adc857a2624ab0d26a2d87baf0 work on ARM64, you should be good to go. But, beware of _other_ projects that may reference those ones and make sure such references are also conditional.

Eilon committed:
Add missing doc comments and make them show up as warnings if missing

javiercn committed:
tmp

javiercn committed:
Use EmbeddedManifestFileProvider and same Blazor Server MSBuild

javiercn committed:
Update project references

pranavkm reviewed (approved):
#yolo

Eilon commented:
I retroactively sign off 😀
